File: solution/207/solution.cpp
struct GrapNode{
    int label;
    std::vector<GrapNode*> neighbors;
    GrapNode(int x):label(x) {};
};

class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        std::vector<GrapNode*> nodes;
        std::vector<int> visits;
        for(int i=0;i<numCourses;i++){nodes.push_back(new GrapNode(i));visits.push_back(-1);}
        for(int i=0;i<prerequisites.size();i++){
            nodes[prerequisites[i].second]->neighbors.push_back(nodes[prerequisites[i].first]);
        }
        for(int i=0;i<nodes.size();i++){
            if(visits[i]==-1&&!DFS(nodes[i], visits)){
                return false;
            }
        }
        return true;
    }
    bool DFS(GrapNode *node, std::vector<int> &visits){
        visits[node->label]=0;
        for(int i=0;i < node->neighbors.size();i++){
            GrapNode* nownode = node->neighbors[i];
            if(visits[nownode->label] == 0)return false;
            else if(visits[nownode->label] == -1&&!DFS(nownode, visits))return false;
        }
        visits[node->label]=1;
        return true;
    }
};

File: solution/207/solution.py
class GrapNode(object):
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        grapNodes = []
        visits = []
        for i in xrange(numCourses):
            grapNodes.append(GrapNode(i))
            visits.append(-1)
        for x in prerequisites:
            grapNodes[x[1]].neighbors.append(grapNodes[x[0]])
        for x in grapNodes:
            if visits[x.label] == -1 and not self.DFS(x, visits):
                return False
        return True

    def DFS(self, node, visits):
        visits[node.label] = 0
        for x in node.neighbors:
            if visits[x.label] == 0:
                return False
            if visits[x.label] == -1 and not self.DFS(x, visits):
                return False
        visits[node.label] = 1
        return True
